File: lib/ts/src/internal/types/VMockCall.ts
import { VTypeDeclaration } from './VTypeDeclaration';
import { ValidationFailure } from '../validation/ValidationFailure';
import { VType } from './VType';
import { validateMockCall } from '../validation/ValidateMockCall';
import { GenerateContext } from '../generation/GenerateContext';
import { generateRandomUMockCall } from '../generation/GenerateRandomUMockCall';
import { ValidateContext } from '../validation/ValidateContext';

export const mockCallName: string = '_ext.Call_';

export class VMockCall extends VType {
    private types: { [key: string]: VType };

    constructor(types: { [key: string]: VType }) {
        super();
        this.types = types;
    }

    public getTypeParameterCount(): number {
        return 0;
    }

    public validate(givenObj: any, typeParameters: VTypeDeclaration[], ctx: ValidateContext): ValidationFailure[] {
        return validateMockCall(givenObj, this.types, ctx);
    }

    public generateRandomValue(
        blueprintValue: any,
        useBlueprintValue: boolean,
        typeParameters: VTypeDeclaration[],
        ctx: GenerateContext,
    ): any {
        return generateRandomUMockCall(this.types, ctx);
    }

    public getName(): string {
        return mockCallName;
    }
}

File: lib/java/src/main/java/io/github/telepact/internal/types/VMockCall.java
package io.github.telepact.internal.types;

import static io.github.telepact.internal.generation.GenerateRandomUMockCall.generateRandomUMockCall;
import static io.github.telepact.internal.validation.ValidateMockCall.validateMockCall;

import java.util.List;
import java.util.Map;

import io.github.telepact.internal.generation.GenerateContext;
import io.github.telepact.internal.validation.ValidateContext;
import io.github.telepact.internal.validation.ValidationFailure;

public class VMockCall implements VType {

    public static final String _MOCK_CALL_NAME = "_ext.Call_";

    public final Map<String, VType> types;

    public VMockCall(Map<String, VType> types) {
        this.types = types;
    }

    @Override
    public int getTypeParameterCount() {
        return 0;
    }

    @Override
    public List<ValidationFailure> validate(Object givenObj,
            List<VTypeDeclaration> typeParameters, ValidateContext ctx) {
        return validateMockCall(givenObj, typeParameters, this.types, ctx);
    }

    @Override
    public Object generateRandomValue(Object blueprintValue, boolean useBlueprintValue,
            List<VTypeDeclaration> typeParameters, GenerateContext ctx) {
        return generateRandomUMockCall(this.types, ctx);
    }

    @Override
    public String getName() {
        return _MOCK_CALL_NAME;
    }

}
